### DISCUSSION TITLE
Are there syntax tree utils for the following to convert back from tree to the source content?

### DISCUSSION THREAD

**Discussion by chrisrzhou** [unifiedjs/unified#98] [Q&A] [answered]
Hi!

In my current `unified-doc` project, I have a way to [convert files](https://github.com/unified-doc/unified-doc/blob/main/packages/unified-doc/lib/file.js#L11) through `hast` trees:
- Start from content
- Always parse to `hast`
- Use `hast` utils to return relevant content (stringified HTML, text content etc).

I thought it might be cool to convert to other file formats, if the ecosystem supports it.  This would be done with simple a few other steps:
- Convert to other syntax trees e.g. `hast-util-to-mdast`, `hast-util-to-xast`, `hast-util-to-nlcst`
- Convert from those syntax tree to its stringified source content.

I notice such utils exist for `hast` and `nlcst` (i.e. `hast-util-to-html`, `nlcst-to-string`) but not for `mdast`, `xast`.  I also notice that `remark-stringify` does this, but I require this as a syntax tree util for reasons mentioned above.

Just wondering if there are any plans to ship this officially within the ecosystem, or any direction pointing to implementing this would be great.  I'm getting excited thinking about how file conversion can be easily done through the unified ecosystem, and look forward to any guidance here!

**Comment by ChristianMurphy** [comment, marked as answer]
> but not for `mdast`, `xast`

what utilities to or from these formats are you looking for?
There are many utils for both: https://github.com/syntax-tree?q=mdast and https://github.com/syntax-tree?q=xast, one may already suit your needs.

specifically https://github.com/syntax-tree/xast-util-to-xml and https://github.com/remarkjs/remark/tree/main/packages/remark-stringify, which stringify their respective ASTs back to text.

**Reply by wooorm** [reply]
For remark-stringify: while it's not a utility per se, the compiler can be access too as it is exported.

In the future, with micromark, I'm thinking of having mdast-util-from-markdown and mdast-util-to-markdown.

**Reply by chrisrzhou** [reply]
@ChristianMurphy `xast-util-to-xml` is what I'm looking for thanks! (I'm probably doing a bad search on `'string'` in the unifiedjs explore search).

@wooorm: YAS. Those two utils are what would solve my problem.  I interface against the `hast` tree and not directly with the compiler at that stage of my code, so having the syntax tree utilities would be ideal.  I might make this step 'hacky' for `mdast` now until the utils are available.

It sounds the all unist syntax trees will eventually have the utilities `<tree>-util-from-<content_type>` and `<tree>-util-to-<content_type>` soon.

**Reply by wooorm** [reply]
https://github.com/syntax-tree/mdast-util-from-markdown

**Reply by chrisrzhou** [reply]
> https://github.com/syntax-tree/mdast-util-from-markdown

Uses `micromark` under the hood too! ❤️ 

I think right now, from `hast`, I have:
- `hast` to `xml` via `hast-util-to-xast + xast-util-to-xml`
- `hast` to `markdown` via `hast-util-to-mdast + X` where `X` is most likely `mdast-util-to-markdown`?

I'm in no rush needing this, but I think it's pretty cool at how the ecosystem is able to close on itself in a pretty well-defined way.  Somewhere in the homepage to illustrate how transformations between content type via their syntax tree representation might be a cool thing to include.  I might contribute and add a tutorial/documentation about it in the near future!

**Reply by wooorm** [reply]
Yeah, think `mdast-util-to-markdown` is next, and not so hard (as the tree is mostly the same, I can mostly copy `remark-stringify`). However, a big thing to figure out is how to do extensions, as they’ll have to hook into all these different things: micromark directly, micromark through mdast-util-from-markdown, mdast-util-to-markdown. And mdast-util-{from,to}-markdown through unified plugins.

> I might contribute and add a tutorial/documentation about it in the near future!

that would be very welcome! We have longer guides and shorter recipes in the [Learn](https://unifiedjs.com/learn/) section.

**Reply by chrisrzhou** [reply]
Thanks @wooorm for the quick mirror implementation of `mdast-util-[]-markdown`!  Only took about 5min and 4 lines of code (technically one line if you ignore the `case` block) to support exporting to `markdown` in `unified-doc`!  Feels great when the ecosystem of micro-utilities come together nicely to do things with very few lines of code!

https://github.com/unified-doc/unified-doc/commit/2b22a1a692ddf799452c15c60701037e9234b009#diff-39296eacbef1de381ecabd4b86df6af7R26-R29

Relating to this, I notice we have `hast-util-to-xast`, `xast-util-to-xml`, `xast-util-from-xml`, but no `xast-util-to-hast`.  Is it perhaps not meaningful for `xast-util-to-hast` to exist given `xast` is very loosely defined and not everything can be meaningfully mapped to hast?

**Reply by wooorm** [reply]
> Is it perhaps not meaningful for xast-util-to-hast to exist given xast is very loosely defined and not everything can be meaningfully mapped to hast?

Yeah, that’s why it didn’t really make sense to me. xast/xml can be lots of non-{html,svg} things, so not sure how to deal with that
